tests/extras/datasets/avro/test_avro_dataset.py

```diff
@@ -0,0 +1,236 @@
+# trademarks of QuantumBlack. The License does not grant you any right or
+# license to the QuantumBlack Trademarks. You may not use the QuantumBlack
+# Trademarks or any confusingly similar mark as a trademark for your product,
+# or use the QuantumBlack Trademarks in any other manner that might cause
+# confusion in the marketplace, including but not limited to in advertising,
+# on websites, or on software.
+#
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from pathlib import PurePosixPath
+
+import pytest
+from fsspec.implementations.http import HTTPFileSystem
+from fsspec.implementations.local import LocalFileSystem
+from gcsfs import GCSFileSystem
+from s3fs.core import S3FileSystem
+
+from kedro.extras.datasets.avro import AVRODataSet
+from kedro.io import DataSetError
+from kedro.io.core import PROTOCOL_DELIMITER, Version
+
+FILENAME = "test.avro"
+
+
+@pytest.fixture
+@pytest.mark.parametrize("tmp_path", ["/tmp"], indirect=True)
+def filepath_avro(tmp_path):
+    return (tmp_path / FILENAME).as_posix()
+
+
+@pytest.fixture
+def dummy_data():
+    return [
+        {"col1": 1, "col2": 3, "col3": 5},
+        {"col1": 2, "col2": 4, "col3": 6},
+    ]
+
+
+@pytest.fixture
+def dummy_schema():
+    return {
+        "namespace": "example.avro",
+        "type": "array",
+        "name": "DataArray",
+        "items": {
+            "type": "record",
+            "name": "DataElement",
+            "fields": [
+                {"name": "col1", "type": "int"},
+                {"name": "col2", "type": "int"},
+                {"name": "col3", "type": "int"},
+            ],
+        },
+    }
+
+
+@pytest.fixture
+def dummy_schema_record():
+    return {
+        "namespace": "example.avro",
+        "type": "record",
+        "name": "DataElement",
+        "fields": [
+            {"name": "col1", "type": "int"},
+            {"name": "col2", "type": "int"},
+            {"name": "col3", "type": "int"},
+        ],
+    }
+
+
+@pytest.fixture
+def dummy_codec():
+    return "null"
+
+
+@pytest.fixture
+def avro_data_set(filepath_avro, dummy_schema):
+    return AVRODataSet(filepath=filepath_avro, schema=dummy_schema,)
+
+
+@pytest.fixture
+def versioned_avro_data_set(
+    filepath_avro, dummy_schema, dummy_codec, load_version, save_version
+):
+    return AVRODataSet(
+        filepath=filepath_avro,
+        schema=dummy_schema,
+        codec=dummy_codec,
+        version=Version(load_version, save_version),
+    )
+
+
+class TestAvroDataSet:
+    def test_credentials_propagated(self, mocker):
+        """Test propagating credentials for connecting to GCS"""
+        mock_fs = mocker.patch("fsspec.filesystem")
+        credentials = {"key": "value"}
+
+        AVRODataSet(filepath=FILENAME, credentials=credentials)
+
+        mock_fs.assert_called_once_with("file", **credentials)
+
+    def test_save_and_load(self, tmp_path, dummy_data, dummy_schema):
```

Comment:
Could you divide this up in several tests? Ideally a unit test would only cover 1 case, so just one happy path or one unhappy path, not a combination of several.